component/client/chat/ClientIndependentChatRoom.tsx

```diff
@@ -22,27 +23,24 @@ class ClientIndependentChatRoom extends React.Component<any, IClientIndependentC
     super();
     this.state = {
       currentUrl: Constants.NOOP_URL,
+      userID: Constants.NOOP_ID,
       username: Constants.NOOP_USERNAME
     };
-    storage.get("username")
-      .then((username) => {
-        if (username) {
+    storage.get(Constants.STORAGE_KEY_USERNAME).then((username) => {
+      storage.get(Constants.STORAGE_KEY_ID).then((userID) => {
+        ClientIndependentChatRoom.getCurrentTabUrl((url) => {
```

Comment:
I don't really like chaining those promises.
Getting userName and getting url are totally separate operations, so they should be done in separate callbacks.
I guess you may try to avoid less calls to setState(). I believe if setState() is called frequently, React would optimize it and consolidate multiple updates to single render. So it is OK to repeatedly call setState(), and only one render() would be triggered if these calls are very close.